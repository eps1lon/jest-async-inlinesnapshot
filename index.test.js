const { toMatchSnapshot, toMatchInlineSnapshot } = require("jest-snapshot");

async function observe(fn) {
  const start = performance.now();
  const result = await fn();

  return {
    // Arbitrary, stable value that observes some global value during `await fn()`.
    slow: performance.now() - start > Number.EPSILON,
    result,
  };
}

expect.extend({
  async toMatchSpeechInlineSnapshotAsync(fn, ...rest) {
    // Jest uses the stacktraces to find the callsite.
    // For async stacktraces it'll point to the wrapped call
    // i.e. every `toMatchSpeechInlineSnapshotAsync` will point to the same call triggering "Jest: Multiple inline snapshots for the same call are not supported."
    this.error = new Error();
    const observation = await observe(fn);
    return toMatchInlineSnapshot.call(this, observation, ...rest);
  },
  toMatchSpeechInlineSnapshot(fn, ...args) {
    const result = fn();
    return toMatchInlineSnapshot.call(this, result, ...args);
  },
  async toMatchSpeechSnapshotAsync(fn, ...args) {
    const observation = await observe(fn);
    return toMatchSnapshot.call(this, observation, ...args);
  },
  toMatchSpeechSnapshot(fn, ...args) {
    const result = fn();
    return toMatchSnapshot.call(this, result, ...args);
  },
});

describe("custom snapshot matcher", () => {
  it("works if sync", () => {
    expect(() => {
      return "sync snapshot #1";
    }).toMatchSpeechSnapshot();

    expect(() => {
      return "sync snapshot #2";
    }).toMatchSpeechSnapshot();
  });

  it("works if async", async () => {
    await expect(async () => {
      return "async snapshot #1";
    }).toMatchSpeechSnapshotAsync();

    await expect(async () => {
      return "async snapshot #2";
    }).toMatchSpeechSnapshotAsync();
  });
});

describe("custom inline snapshot matcher", () => {
  it("works if sync", () => {
    expect(() => {
      return "sync snapshot #1";
    }).toMatchSpeechInlineSnapshot(`"sync snapshot #1"`);

    expect(() => {
      return "sync snapshot #2";
    }).toMatchSpeechInlineSnapshot(`"sync snapshot #2"`);
  });

  it("works if async", async () => {
    await expect(async () => {
      return "async inline snapshot #1";
    }).toMatchSpeechInlineSnapshotAsync();
    await expect(async () => {
      return "async inline snapshot #2";
    }).toMatchSpeechInlineSnapshotAsync(`
            Object {
              "result": "async inline snapshot #?",
              "slow": true,
            }
          `);
  });
});
