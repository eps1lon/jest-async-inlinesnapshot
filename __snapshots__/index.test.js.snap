// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`custom snapshot matcher works if async 1`] = `
Object {
  "result": "async snapshot #1",
  "slow": true,
}
`;

exports[`custom snapshot matcher works if async 2`] = `
Object {
  "result": "async snapshot #2",
  "slow": true,
}
`;

exports[`custom snapshot matcher works if sync 1`] = `"sync snapshot #1"`;

exports[`custom snapshot matcher works if sync 2`] = `"sync snapshot #2"`;
